### PR TITLE
Improvement of LanguageError Lexeme Usage

### LINKED ISSUE #819: 📋 [TASK] - Improvement of LanguageError Lexeme Usage
### Parent Feature

#914 

### Task Description

As an AaC Developer, I want to implement the identified improvements for the LanguageError usage of the Lexemes, so that I provide a reliable and useful tool for end users.

### Acceptance Criteria

- [ ] LanguageError Lexeme usage is modified as necessary to better support utilization. 
  - [ ] get_location_str method is modified to return SourceLocation
  - [ ] DefinitionParser class is updated to use modified get_location method
  - [ ] Accurate location data from an erroneous parsing is provided to ExecutionMessage
- [ ] Unit tests for the Lexemes provided to LanguageError are updated.
- [ ] Documentation about utilizing Lexemes is expanded to cover the LangaugeError Lexeme usage layer. 

### Unresolved questions

_No response_

### PR DESCRIPTION
**Pull request by bunchw**
# Description

Improvement of LanguageError Lexeme Usage

# Linked Items:

Closes/Fixes/Resolves #819 

### Added

init and str methods to language error, and a language error test class

### Changed

- _Describe any changes in existing functionality._

### Deprecated

- _Describe any deprecated features._

### Removed

- _Describe any removed features._

### Fixed

- _Describe any bug fixes._

### Security

- _Describe any security-related changes._

# Checklist:

- [ ] I updated project documentation to reflect my changes.
- [x] My changes generate no new warnings.
- [x] I updated new and existing unit tests to account for my changes.
- [x] I linked the associated item(s) to be closed.
- [x] I bumped the version.
- [x] I added the labels corresponding to my changes.

### CODE REVIEW

File: python/src/aac/context/language_error.py [hunk 1]
```diff
@@ -8,3 +8,23 @@ class LanguageError(Exception):
...
+        An init constructor for the Language Error Class.
+
+        Args:
+            message(str): The error message.
+            location(str): The source location of the error.
+        """
+        self.message = message
+        self.location = location
+
+    def __str__(self) -> str:
```
Comment (lizzcondrey) [inline python/src/aac/context/language_error.py:23]: While this isn't an incorrect way to set this, if we are wanting it to be a way of fetching a formatted string of the LanguageError content, you would not want to house it in a "protected" method like this with the name mangling. You would have it as a public method without the name mangling (so no surrounding underscores). That way people would understand it is meant to be used and is not just some internal representation. If you are intending for it to only be used by internal logic and not an exposed usable method for end users, then you would make it "private" and only have a preceding underscore to the start of the name, i.e. _get_language_error_as_str. So, we just need to decide how this is wanting to be used and adjust the method name accordingly.

File: python/src/aac/context/language_error.py [hunk 2]
```diff
@@ -8,3 +8,23 @@ class LanguageError(Exception):
...
+    def __init__(self, message: str, location: str):
+        """
+        An init constructor for the Language Error Class.
+
+        Args:
+            message(str): The error message.
+            location(str): The source location of the error.
+        """
+        self.message = message
+        self.location = location
```
Comment (lizzcondrey) [inline python/src/aac/context/language_error.py:21]: Should still provide the type hints here. 
```suggestion
        self.message: str = message
        self.location: Optional[str] = location
```

Comment (lizzcondrey) [reply python/src/aac/context/language_error.py:21]: Though, we need to discuss if we really want the location portion to be an optional attribute for the error. That is what allows for us to pass `None` in the `location` entries of a `LanguageError`. Do we really want a `None` or do we want a paired down `location` of at minimum returning the file that triggered the error?

File: python/tests/test_aac/context/test_language_error.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+from unittest import TestCase
+
+from aac.context.language_context import LanguageContext
+from aac.context.language_error import LanguageError
+
+class TestLanguageError(TestCase):
+    def test_language_error(self):
```
Comment (lizzcondrey) [inline python/tests/test_aac/context/test_language_error.py:7]: With having the `location` set as an optional, should also test that it handles fine receiving `None`, if we don't decide to move away from that implementation.

**Review by lizzcondrey — CHANGES_REQUESTED** [review]
Make sure to bump version.

File: python/src/aac/context/language_error.py (hunk 1 above)
Comment (bunchw) [reply python/src/aac/context/language_error.py:23]: __str__ is a special python method, like __init__.  It wont be called.  __str__ is basically whats called when you print() something, it just formats the class into a readable format.

File: python/src/aac/context/language_error.py (hunk 2 above)
Comment (bunchw) [reply python/src/aac/context/language_error.py:21]: If we have None as location, we could instead give it a string like "No Location" or something, and remove the optional from the type.

File: python/src/aac/context/language_error.py (hunk 1 above)
Comment (lizzcondrey) [reply python/src/aac/context/language_error.py:23]: Correct. So, this was not added with the intent of being used elsewhere within the code, but just to leverage that functionality?

Comment (bunchw) [reply python/src/aac/context/language_error.py:23]: Yes.

File: python/src/aac/context/language_error.py (hunk 2 above)
Comment (bunchw) [reply python/src/aac/context/language_error.py:21]: Also, I dont think it needs the type hints because its not creating the class variable, its just setting them.  It has type hints above in the class body.

Comment (lizzcondrey) [reply python/src/aac/context/language_error.py:21]: > If we have None as location, we could instead give it a string like "No Location" or something, and remove the optional from the type.

So, meaning where we are actually raising the errors if no applicable location provide the "No Location" notice instead?

File: python/src/aac/context/language_error.py [hunk 4]
```diff
@@ -8,3 +8,23 @@ class LanguageError(Exception):
...
     message: str = attrib(validator=validators.instance_of(str))
     location: Optional[str] = attrib(validator=validators.optional(validators.instance_of(str)))
+
+    def __init__(self, message: str, location: str):
+        """
+        An init constructor for the Language Error Class.
+
+        Args:
+            message(str): The error message.
+            location(str): The source location of the error.
```
Comment (lizzcondrey) [inline python/src/aac/context/language_error.py:18]:
```suggestion
            message (str): The error message.
            location (str): The source location of the error.
```

File: python/src/aac/context/language_error.py [hunk 5]
```diff
@@ -8,3 +8,23 @@ class LanguageError(Exception):
     """A base class representing a language error condition."""
```
Comment (lizzcondrey) [inline python/src/aac/context/language_error.py:8]:
```suggestion
    """A base class representing a language error."""
```

**Review by lizzcondrey — CHANGES_REQUESTED** [review]

File: python/src/aac/execute/command_line.py [hunk 6]
```diff
@@ -86,7 +87,7 @@ def wrapper(*args, **kwargs):
                 plugin_name,
                 "exception",
                 ExecutionStatus.GENERAL_FAILURE,
-                [ExecutionMessage(e.message, MessageLevel.ERROR, None, e.location)],
+                [ExecutionMessage(e.message, MessageLevel.ERROR, AaCFile(e.location, True, True), None)],
```
[the remark this answers is not in the document]
Comment (bunchw) [reply python/src/aac/execute/command_line.py:90]: The fourth argument for execution message is a SourceLocation, the column and row, not the file.  e.location is a string containing the name of the file.

File: python/src/aac/plugins/check/check_aac_impl.py [hunk 7]
```diff
@@ -137,8 +137,8 @@ def check_schema_constraint(
                     # if the field is a list, check each item in the list
                     if type(getattr(check_me, field.name)) != list:
                         raise LanguageError(
-                            f"Value of '{field.name}' was expected to be list, but was '{type(getattr(check_me, field.name))}'. File: '{source_definition.source.uri}'",
-                            None
+                            f"Value of '{field.name}' was expected to be list, but was '{type(getattr(check_me, field.name))}'",
+                            f'{source_definition.source.uri}'
```
Comment (lizzcondrey) [inline python/src/aac/plugins/check/check_aac_impl.py:141]: Wouldn't drop the additional file designator. Always best to be as explicit in returned errors as you can be.

```suggestion
                            f"Encountered in file: {source_definition.source.uri}"
```

File: python/tests/test_aac/context/test_language_error.py (hunk 3 above)
Comment (lizzcondrey) [reply python/tests/test_aac/context/test_language_error.py:7]: This can be ignored since we are going to try and move away from having it be an optional and going to force some kind of location provision, even if a notice of something like "No location"

File: python/src/aac/context/lexeme.py [hunk 8]
```diff
@@ -20,6 +20,19 @@ class Lexeme:
     source: str = attrib(validator=validators.instance_of(str))
     value: str = attrib(validator=validators.instance_of(str))
 
+    def __init__(self, location: SourceLocation, source: str, value: str):
+        """
+        An init constructor for the Language Error Class.
```
Comment (lizzcondrey) [inline python/src/aac/context/lexeme.py:25]:
```suggestion
        An init constructor for the Lexeme Class.
```

File: python/src/aac/context/lexeme.py [hunk 9]
```diff
@@ -20,6 +20,19 @@ class Lexeme:
     source: str = attrib(validator=validators.instance_of(str))
     value: str = attrib(validator=validators.instance_of(str))
 
+    def __init__(self, location: SourceLocation, source: str, value: str):
```
Comment (lizzcondrey) [inline python/src/aac/context/lexeme.py:23]: So, was similar behavior to what was being seen with `LanguageError` found to be happening with the `Lexeme` class then?

**Review by lizzcondrey — CHANGES_REQUESTED** [review]

File: python/src/aac/execute/command_line.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (lizzcondrey) [reply python/src/aac/execute/command_line.py:90]: Going to pull that info and make sure it is included with the `ExecutionMessage` task as part of the evaluations and potential changes.

File: python/src/aac/context/lexeme.py (hunk 9 above)
Comment (bunchw) [reply python/src/aac/context/lexeme.py:23]: I didn't see anywhere that had that issue, but it's never a bad idea to have an __init__ method anyway, and it could avoid a problem down the line.

File: python/src/aac/plugins/check/check_aac_impl.py (hunk 7 above)
Comment (bunchw) [reply python/src/aac/plugins/check/check_aac_impl.py:141]: Thats what the __str__ method is for.  Its not neccessary to have a designator before the location because the __str__ has one, and when the error is diplayed it calls the __str__ method to display it.  It will show as 
"Error: {message}. Location: {location}"

Comment (bunchw) [reply python/src/aac/plugins/check/check_aac_impl.py:141]: We can change the actual text of the __str__ method to be anything if we want to also.

Comment (lizzcondrey) [reply python/src/aac/plugins/check/check_aac_impl.py:141]: Okay, so that is part of what I was trying to get at when trying to understand why you implemented the formatting method how you did. I'll try to take a look for making sure this prints out like you expected. Because as it's leveraged right now, the `LanguageError` is just caught as Exceptions then the `message` and `location` portions are accessed from the exception. So, your formatting method won't really play a role as it stands.

Comment (lizzcondrey) [reply python/src/aac/plugins/check/check_aac_impl.py:141]: Is that something you are wanting to change with how the `LanguageError` is leveraged by pieces like `ExecutionMessage` to make use of the `__str__` formatting method you added?

Comment (bunchw) [reply python/src/aac/plugins/check/check_aac_impl.py:141]: I don't think anything would have to change about ExecutionMessage, it doesnt directly interact with language error.  In some places, e.message and e.location are passed into an ExecutionMessage, and that will work as it does currently.

Comment (lizzcondrey) [reply python/src/aac/plugins/check/check_aac_impl.py:141]: So, I am only finding LanguageError to be caught then the individual pieces referenced for displaying the content. Because of that, I don't think your formatting method is ever going to be used. Have you seen different behavior somewhere that I am missing?

Comment (bunchw) [reply python/src/aac/plugins/check/check_aac_impl.py:141]: When a language error is not caught and displays in the terminal, the __str__ method is called.

Comment (lizzcondrey) [reply python/src/aac/plugins/check/check_aac_impl.py:141]: Had side bar further discussing and reviewing the format method and its usage by the tool as backup provision when the exception is just raised and not caught for further control. Should be good to move on from this now.

**Review by lizzcondrey — APPROVED** [review]
